lib/accesslists/collection.go

```diff
@@ -0,0 +1,151 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package accesslists
+
+import (
+	"context"
+	"slices"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types/accesslist"
+)
+
+// Collection represents a complete batch of access from one source (e.g. entraID or okta)
+// That implies that all access lists and members in the batch are expected to be created/updated together
+// because snapshot have a final complete state of the access lists and their members hierarchy.
+// For instance imagine an access list import from EntraID directory where  the access list
+// can't be modified outside EntraID and teleport only fetches the complete state of access lists and members from EntraID
+// This is useful to validate nested hierarchy on the fly and preset the reference updates (Status.MemberOf and Status.OwnerOf)
+// before starting the actual creation/update process in the backend.
+// Where the relation between access list outside the Collection is not possible because collection source are
+// aware only of the access lists and members they contain.
+type Collection struct {
+	// MembersByAccessList maps access list names to their members
+	MembersByAccessList map[string][]*accesslist.AccessListMember
+	// AccessListMap is an internal map for fast lookup by name
+	AccessListMap map[string]*accesslist.AccessList
```

Comment:
can both of those fields be private? Or do they need to be accessed externally?

Will this struct be used by concurrent readers/writers? Is there a need for a lock?

Reply:
The types are not tread safe and the struct is designed to work with sequential flow. 

>  Or do they need to be accessed externally? 

It will be accessed externally  .